### ISSUE TITLE
Spring 3.1 EntitySpecification + Pageable + EntityGraph Pages In-Memory

### ISSUE THREAD

**Issue by thejeff77** [spring-projects/spring-data-jpa#3206]
This is both a question and/or a bug report and/or a feature request... depending on how this conversation goes. I could use some help with any ways to accomplish what I'm trying to do.

(kotlin repo function:)
```
@Repository
interface PuppyJpaRepository :
    JpaRepository<PuppyEntity, UUID> {
    @EntityGraph("PuppyEntity.subgraph")
    override fun findAll(spec: Specification<PuppyEntity>, pageable: Pageable): Page<PuppyEntity>
}
```

I have the following setting enabled for performance reasons: `spring.jpa.properties.hibernate.query.fail_on_pagination_over_collection_fetch = true`

So I get this error:  `firstResult/maxResults specified with collection fetch. In memory pagination was about to be applied.`

How do I accomplish Paging, or even simply a limit query with an EntitySpecification, and an EntityGraph?

What I've tried:
- Tried finding ways to use .setMaxResults() on the EntitySpecification, unfortunately I can't find a way to do this without access to the entityManager which is abstracted away via spring data repositories. Is there a way to do this?
- Tried keeping the Pageable very simple - I.E Page 0 with a size to see if it was smart enough to setMaxResults.. it didn't work
- Tried looking into BlazePersistence, unfortunately it requires creation of views which is heavier than I'd like to get into. It also seems to have issues with kotlin.
- Looked into separating the count query from the main query. This is also heavy and has a performance impact (although less so than the N+1 issue). [Example article for this, but it doesn't use EntitySpecification](https://www.appsloveworld.com/springboot/100/25/spring-data-jpa-pagination-hhh000104).

Thanks for any help/info! Is there a way to set a limit for such an EntitySpecification / EntityGraph simply with spring data JPA?

Thanks for any help / comments / questions / context!

**Comment by thejeff77** [comment]
Specification<PuppyEntity> = Entity Specification.. sorry I thought that was a class name so I didn't put a space in-between. 

I'd assume that setting max results on such a query would manifest as a limit query, which would definitely not have to pull everything beyond the limit into memory. Would that be a correct assumption?

I can't call .setMaxResults() since this is a Spring data abstraction layer.. I have to pass in Pageable which in this case pages in memory.

Please correct any of my incorrect assumptions above.

Basically I'd like to limit the results without loading the whole thing into memory.

**Comment by mp911de** [comment]
The in-memory aspect is a consequence of joining through `@EntityGraph`. Hibernate evaluates it and I don't think there is anything more that you can set. Clearly, you can pass any query hints using annotations. `setMaxResults(…)` is a consequence of providing a `PageRequest` object.

Therefore, we cannot proceed here. If you find something in the future that would allow more fine-grained control, we're happy to reopen the ticket.

**Comment by thejeff77** [comment]
@mp911de I guess I'm wondering why spring data won't use setMaxResults alone when the page request is for page number 0 (I.E. the first page), and not page in-memory for this case. 

This would allow a limit to be set to a query that isn't technically paging yet, as page isn't > 0. 

I can't set a queryhint dynamically in spring data, so being able to setMaxResults on this "clearly" isn't possible unless it was a static hard coded value. 

To me it seems like my question is more turning into a bug report / optimization feature request?

Am I making sense or am I missing something?

**Comment by mp911de** [comment]
> I'm wondering why spring data won't use setMaxResults alone when the page request is for page number 0 (I.E. the first page)

Care to elaborate how you identified that `Query.setMaxResults(…)` isn't invoked?

With a reproducer of:

```java

interface MyEntityRepository extends CrudRepository<MyEntity, Long>, JpaSpecificationExecutor<MyEntity> {…}

MyEntityRepository repo = …
repo.findAll(Specification.where((root, query, criteriaBuilder) -> {
	return criteriaBuilder.equal(criteriaBuilder.literal(1), criteriaBuilder.literal(1));
}), Pageable.ofSize(25));
```

I get the confirmation through Hibernate logging that the limit is being considered:

```
Hibernate: select m1_0.id,m1_0.name from my_entity m1_0 where 1=1 offset ? rows fetch first ? rows only
2023-11-09T08:56:39.770+01:00 TRACE 89336 --- [           main] org.hibernate.orm.jdbc.bind              : binding parameter [1] as [INTEGER] - [0]
2023-11-09T08:56:39.770+01:00 TRACE 89336 --- [           main] org.hibernate.orm.jdbc.bind              : binding parameter [2] as [INTEGER] - [25]
```

**Comment by thejeff77** [comment]
cc @mp911de
